shell/platform/android/io/flutter/embedding/engine/android/AndroidKeyProcessor.java

```diff
@@ -0,0 +1,87 @@
+package io.flutter.embedding.engine.android;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.view.KeyCharacterMap;
+import android.view.KeyEvent;
+
+import io.flutter.embedding.engine.systemchannels.KeyEventChannel;
+
+public class AndroidKeyProcessor {
+  private final KeyEventChannel keyEventChannel;
+  private int combiningCharacter;
+
+  public AndroidKeyProcessor(@NonNull KeyEventChannel keyEventChannel) {
+    this.keyEventChannel = keyEventChannel;
+  }
+
+  public void onKeyUp(@NonNull KeyEvent keyEvent) {
+    Character complexCharacter = applyCombiningCharacterToBaseCharacter(keyEvent.getUnicodeChar());
+    keyEventChannel.keyUp(
+        new KeyEventChannel.FlutterKeyEvent(keyEvent, complexCharacter)
+    );
+  }
+
+  public void onKeyDown(@NonNull KeyEvent keyEvent) {
+    Character complexCharacter = applyCombiningCharacterToBaseCharacter(keyEvent.getUnicodeChar());
+    keyEventChannel.keyDown(
+        new KeyEventChannel.FlutterKeyEvent(keyEvent, complexCharacter)
+    );
+  }
+
+  /**
+   * Applies the given Unicode character in {@code newCharacterCodePoint} to a previously
+   * entered Unicode combining character and returns the combination of these characters
+   * if a combination exists.
+   * <p>
+   * This method mutates {@link #combiningCharacter} over time to combine characters.
+   * <p>
+   * One of the following things happens in this method:
+   * <ul>
+   *   <li>If no previous {@link #combiningCharacter} exists and the {@code newCharacterCodePoint}
+   *   is not a combining character, then nothing happens and {@code newCharacterCodePoint} is returned.</li>
```

Comment:
This is definitely a nit, but:
```suggestion
   *   is not a combining character, then {@code newCharacterCodePoint} is returned.</li>
```

Reply:
Done